// ajv@6.3.2/lib/vocabularies/jtd/type.js
(function (deps, factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(deps, factory);
    }
})(["require", "exports", "../../compile/codegen", "../../runtime/timestamp", "../../compile/util", "./metadata", "./error"], function (require, exports) {
    {
        CodeKeywordDefinition, KeywordErrorDefinition;
    }
    from;
    "../../types";
    {
        KeywordCxt;
    }
    from;
    "../../compile/validate";
    var codegen_1 = require("../../compile/codegen");
    var timestamp_1 = require("../../runtime/timestamp");
    var util_1 = require("../../compile/util");
    var metadata_1 = require("./metadata");
    var error_1 = require("./error");
    "type", JTDType, JTDType >
    ;
    "int8" | "uint8" | "int16" | "uint16" | "int32" | "uint32";
    exports.intRange = {
        int8: [-128, 127, 3],
        uint8: [0, 255, 3],
        int16: [-32768, 32767, 5],
        uint16: [0, 65535, 5],
        int32: [-2147483648, 2147483647, 10],
        uint32: [0, 4294967295, 10],
    };
    "boolean" | "string" | "timestamp" | "float32" | "float64" | IntType;
    var error = {
        message: function (cxt) { return error_1.typeErrorMessage(cxt, cxt.schema); },
        params: function (cxt) { return error_1.typeErrorParams(cxt, cxt.schema); },
    };
    var def = {
        keyword: "type",
        schemaType: "string",
        error: error,
        code: function (cxt) {
            metadata_1.checkMetadata(cxt);
            var gen = cxt.gen, data = cxt.data, schema = cxt.schema, parentSchema = cxt.parentSchema;
            var cond;
            switch (schema) {
                case "boolean":
                case "string":
                    cond = (_a = ["typeof ", " == ", ""], _a.raw = ["typeof ", " == ", ""], codegen_1._(_a, data, schema));
                    break;
                case "timestamp": {
                    var vts = util_1.useFunc(gen, timestamp_1.default);
                    cond = (_b = ["", " instanceof Date || (typeof ", " == \"string\" && ", "(", "))"], _b.raw = ["", " instanceof Date || (typeof ", " == \"string\" && ", "(", "))"], codegen_1._(_b, data, data, vts, data));
                    break;
                }
                case "float32":
                case "float64":
                    cond = (_c = ["typeof ", " == \"number\""], _c.raw = ["typeof ", " == \"number\""], codegen_1._(_c, data));
                    break;
                default: {
                    var _d = exports.intRange[schema], min = _d[0], max = _d[1], as = IntType, cond_1 = (_e = ["typeof ", " == \"number\" && isFinite(", ") && ", " >= ", " && ", " <= ", " && !(", " % 1)"], _e.raw = ["typeof ", " == \"number\" && isFinite(", ") && ", " >= ", " && ", " <= ", " && !(", " % 1)"], codegen_1._(_e, data, data, data, min, data, max, data));
                }
            }
            cxt.pass(parentSchema.nullable ? codegen_1.or((_f = ["", " === null"], _f.raw = ["", " === null"], codegen_1._(_f, data)), cond) : cond);
            var _a, _b, _c, _e, _f;
        },
    };
    exports.default = def;
});
