// ajv@6.3.2/lib/runtime/validation_error.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
(function (deps, factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(deps, factory);
    }
})(["require", "exports"], function (require, exports) {
    {
        ErrorObject;
    }
    from;
    "../types";
    var ValidationError = (function (_super) {
        __extends(ValidationError, _super);
        function ValidationError(errors) {
            _super.call(this, "validation failed");
            this.readonly = errors;
            this.readonly = ajv;
            this.readonly = validation;
            this.errors = errors;
            this.ajv = this.validation = true;
        }
        ValidationError.prototype.Partial = ;
        return ValidationError;
    })(Error);
    exports.default = ValidationError;
});
